src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -115,19 +117,139 @@ public sealed override object ReadInline(object item, IEdmTypeReference edmType,
+                {
+                    yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                }
+            }
+            else
+            {
+                foreach (ODataItemBase odataItemBase in resourceSet.Resources)
+                {
+                    ODataResourceWrapper resourceWrapper = odataItemBase as ODataResourceWrapper;
+
+                   if(resourceWrapper != null)
+                    {
+                        ODataDeletedResource deletedResource = resourceWrapper.ResourceBase as ODataDeletedResource;
+
+                        if (deletedResource != null)
+                        {
+                            yield return DeSerializeDeletedEntity(readContext, deletedResource);
+                        }
+                        else
+                        {
+                            yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                        }
+                    }
+                    else
+                    {
+                        ODataDeltaLinkWrapper deltaLinkWrapper = odataItemBase as ODataDeltaLinkWrapper;
+
+                        Contract.Assert(deltaLinkWrapper != null, "ODataDeltaLinkWrapper should not be null");
+
+                        yield return DeserializeDeltaLink(readContext, deltaLinkWrapper, resourceSet.ResourceSetBase.TypeName);                      
+                    }
+                    
+                }
+            }            
+        }
+
+
+        private static IEdmDeltaLinkBase DeserializeDeltaLink(ODataDeserializerContext readContext, ODataDeltaLinkWrapper deltaLinkWrapper, string typeName)
+        {
+            ODataDeltaLinkBase deltalink = deltaLinkWrapper.DeltaLink;
+
+            if (deltalink == null)
+            {
+                throw new ODataException("Deleted link not present");
+            }            
+
+            IEdmModel model = readContext.Model;
+
+            if (model == null)
+            {
+                throw Error.Argument("readContext", SRResources.ModelMissingFromReadContext);
+            }
+
+            IEdmStructuredType actualType = model.FindType(typeName) as IEdmStructuredType;
+            if (actualType == null)
+            {
+                throw new ODataException(Error.Format(SRResources.ResourceTypeNotInModel, typeName));
+            }
 
-            foreach (ODataResourceWrapper resourceWrapper in resourceSet.Resources)
+            if (actualType.IsAbstract)
             {
-                yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                string message = Error.Format(SRResources.CannotInstantiateAbstractResourceType, deltalink.TypeAnnotation.TypeName);
+                throw new ODataException(message);
             }
+
+            IEdmEntityType actualEntityType = actualType as IEdmEntityType;
```

Comment:
>actualEntityType [](start = 27, length = 16)

maybe add assert that actualEdmEntityType is not null?  or, better yet, on line 185, just cast it to IEdmEntityType and let the check on 186 handle the null case.